varinfo/cf_config.py

```diff
@@ -160,40 +155,55 @@ def _create_attributes_object(cf_item: dict) -> dict[str, str]:
 
-        """
-        if variable is not None:
-            cf_overrides = self._get_matching_attributes(self._cf_overrides, variable)
-            cf_supplements = self._get_matching_attributes(
-                self._cf_supplements, variable
-            )
-        else:
-            cf_overrides = self._cf_overrides
-            cf_supplements = self._cf_supplements
+        First iterate through the self.cf_overrides and find all items with a
+        variable pattern that matches the supplied variable (or group) path.
 
-        return {'cf_overrides': cf_overrides, 'cf_supplements': cf_supplements}
+        Next sort that dictionary, so that matching patterns are:
 
-    @staticmethod
-    def _get_matching_attributes(
-        cf_references: dict[str, dict[str, str]], variable: str
-    ) -> dict[str, str]:
-        """Iterate through either the self._cf_supplements or
-        self._cf_overrides and extract a dictionary that combines all
-        applicable attributes that apply to the specified variable. If
-        there are conflicting values for the same attribute, only the last
-        value will be returned for that attribute.
+        * Primarily sorted from shallowed to deepest, by counting the number of
+          slashes in the string.
+        * Within each depth (with the same number of slashes), patterns are
+          sorted from shortest to longest string length.
+
+        It is assumed that regular expressions that match deeper elements of
+        a file hierarchy are intended to be more specifically applied, and that
+        within a given depth, the string length is a proxy for specificity.
+
+        Last, combine the attribute names and values from each matching
+        override item. Because of the ordering in the previous step, if there
+        are multiple values supplied for the same metadata attribute, the value
+        retained will be the one with the longest variable pattern, which is a
+        proxy for how specific the override is.
 
         """
-        references = {}
-        for pattern, attributes in cf_references.items():
-            if re.match(pattern, variable) is not None:
-                references.update(attributes)
+        matching_overrides = {
+            pattern: attributes
+            for pattern, attributes in self.cf_overrides.items()
+            if re.match(pattern, variable_path) is not None
+        }
+
+        # Order override items by:
+        # First: Depth of the variable hierarchy included in the regular
+        # expression pattern (number of slashes), shallowest to deepest.
+        # Second: The length of the variable pattern, from shorted to longest.
+        # The second ordering is applied in to regular expressions specifying
+        # the same depth of variable hierarchy.
+        sorted_overrides = dict(
+            sorted(
+                matching_overrides.items(),
+                key=lambda pattern: (pattern[0].count('/'), len(pattern[0])),
```

Comment:
This is probably worth a separate comment - the `sorted` function can have a tuple returned from the key function. The documentation isn't super clear, but it seems that the ordering is done primarily on the first element of that tuple, followed by the second element. (I've beefed up the tests a bit, and things seem to be going well in those)